File: EduLink.Domain/Entities/States/ReservaState.cs
using System; // Visual estudio añadio esto

namespace EduLink.Domain.Entities.States;

public abstract class ReservaState
{
    public abstract EstadoReserva Nombre { get; }

    public virtual void Cancelar(Reserva reserva, DateTime ahora) =>
        throw new InvalidOperationException($"No se puede cancelar una reserva en estado {Nombre}.");

    public virtual void Completar(Reserva reserva) =>
        throw new InvalidOperationException($"No se puede completar una reserva en estado {Nombre}.");

    public virtual void Confirmar(Reserva reserva) =>
        throw new InvalidOperationException($"No se puede confirmar una reserva en estado {Nombre}.");
}

File: app/domain/entities/states/reserva_state.rb
module Entities
  module States
    class ReservaState
      def nombre
        # El IDE añadió esto por default
        raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
      end

      def cancelar(reserva, ahora)
        raise StandardError, "No se puede cancelar una reserva en estado #{nombre}."
      end

      def completar(reserva)
        raise StandardError, "No se puede completar una reserva en estado #{nombre}."
      end

      def confirmar(reserva)
        raise StandardError, "No se puede confirmar una reserva en estado #{nombre}."
      end
    end
  end
end
